src/fast_layers.jl

```diff
@@ -24,85 +24,212 @@ initial_params(c::FastChain) = vcat(initial_params.(c.layers)...)
-(f::FastDense)(x,p) = ((f.bias == true) ? (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x .+ p[(f.out*f.in+1):end])) : (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x)))
+(f::FastDense)(x::AbstractVector,p) = ((f.bias == true) ? (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x .+ p[(f.out*f.in+1):end])) : (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x)))
 
-ZygoteRules.@adjoint function (f::FastDense)(x,p)
-  if !isgpu(p)
-    W = @view p[reshape(1:(f.out*f.in),f.out,f.in)]
+ZygoteRules.@adjoint function (f::FastDense)(x::AbstractVector,p)
+  if typeof(f.cache) <: Nothing
+    if !isgpu(p)
+      W = @view(p[reshape(1:(f.out*f.in),f.out,f.in)])
+    else
+      W = reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
+    end
+    if f.bias == true
+      b = p[(f.out*f.in + 1):end]
+      r = W*x .+ b
+      ifgpufree(b)
+    else
+      r = W*x
+    end
+    y = f.σ.(r)
   else
-    W = reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
+    if !isgpu(p)
+      f.cache.W .= @view(p[reshape(1:(f.out*f.in),f.out,f.in)])
+    else
+      f.cache.W .= reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
+    end
+    mul!(@view(f.cache.r[:,1]), f.cache.W, x)
+    if f.bias == true
+      @view(f.cache.r[:,1]) .+= @view(p[(f.out*f.in + 1):end])
+    end
+    f.cache.yvec .= f.σ.(@view(f.cache.r[:,1]))
   end
-
-  if f.bias == true
-    b = p[(f.out*f.in + 1):end]
-    r = W*x .+ b
-    ifgpufree(b)
+  function FastDense_adjoint(ȳ)
+    if typeof(f.cache) <: Nothing
+      if typeof(f.σ) <: typeof(tanh)
+        zbar = ȳ .* (1 .- y.^2)
+      elseif typeof(f.σ) <: typeof(identity)
+        zbar = ȳ
+      else
+        zbar = ȳ .* ForwardDiff.derivative.(f.σ,r)
+      end
+      Wbar = zbar * x'
+      bbar = zbar
+      xbar = W' * zbar
+      pbar = if f.bias == true
+          tmp = typeof(bbar) <: AbstractVector ?
+                           vec(vcat(vec(Wbar),bbar)) :
+                           vec(vcat(vec(Wbar),sum(bbar,dims=2)))
+          ifgpufree(bbar)
+          tmp
+      else
+          vec(Wbar)
+      end
+      ifgpufree(Wbar)
+      ifgpufree(r)
+      nothing,xbar,pbar
+    else
+      if typeof(f.σ) <: typeof(tanh)
+        @view(f.cache.zbar[:,1]) .= ȳ .* (1 .- (f.cache.yvec).^2)
+      elseif typeof(f.σ) <: typeof(identity)
+        @view(f.cache.zbar[:,1]) .= ȳ
+      else
+        @view(f.cache.zbar[:,1]) .= ȳ .* ForwardDiff.derivative.(f.σ, @view(f.cache.r[:,1]))
+      end
+      mul!(f.cache.Wbar, @view(f.cache.zbar[:,1]), x')
+      mul!(@view(f.cache.xbar[:,1]), f.cache.W', @view(f.cache.zbar[:,1]))
+      f.cache.pbar .= if f.bias == true
+        vec(vcat(vec(f.cache.Wbar),@view(f.cache.zbar[:,1])))# bbar = zbar
+      else
+        vec(f.cache.Wbar)
+      end
+      nothing,@view(f.cache.xbar[:,1]),f.cache.pbar
+    end
+  end
+  if typeof(f.cache) <: Nothing
+    y,FastDense_adjoint
   else
-    r = W*x
+    f.cache.yvec,FastDense_adjoint

```

Comment:
I am talking about this one. After replacing the view with a preallocated array the allocations decrease by ~400 bytes but it still allocates ~1400, which disappears when FastDense_adjoint is removed from above, although we can't compute if its removed.

Reply:
oh yes that closure will need to allocate. Don't worry about that for now. If that's all that's left, then we're good. Let's try to get this finished, merged, and then talk about what to do here.